pkg/rbd/nodeserver.go

```diff
@@ -538,6 +539,63 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 	return &csi.NodeUnstageVolumeResponse{}, nil
 }
 
+func (ns *NodeServer) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeRequest) (*csi.NodeExpandVolumeResponse, error) {
+
+	volumeID := req.GetVolumeId()
+	if volumeID == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeExpandVolume volume ID must be provided")
+	}
+	volumePath := req.GetVolumePath()
+	if volumePath == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeExpandVolume volume path must be provided")
+	}
+
+	if acquired := ns.VolumeLocks.TryAcquire(volumeID); !acquired {
+		klog.Infof(util.Log(ctx, util.VolumeOperationAlreadyExistsFmt), volumeID)
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volumeID)
+	}
+	defer ns.VolumeLocks.Release(volumeID)
+
+	_, err := getVolumeName(volumeID)
+	if err != nil {
+		// error ErrInvalidVolID may mean this is an 1.0.0 version volume, check for name
+		// pattern match in addition to error to ensure this is a likely v1.0.0 volume
+		if _, ok := err.(ErrInvalidVolID); !ok || !isLegacyVolumeID(volumeID) {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+
+	diskMounter := &mount.SafeFormatAndMount{Interface: ns.mounter, Exec: mount.NewOsExec()}
+	volumePath = volumePath + "/" + req.GetVolumeId()
```

Comment:
Done.

Reply:
Done.

Reply:
Right now the way the code is modified can possibly lead to a vulnerability. We need to ensure we are working on the Staging path, and not open the image meta stash without checking the same.

If the publish path were sent in here, and the user of the PV created a file that matched another image stash file, we would end up resizing the same.

As checks we need to ensure the passed in path is NOT a mount point (this eliminates it being the publish path), and if found to be a mount point then error out, as we cannot reach the stash file.

Further, if the passed in path is not a mount point, then we can ensure the passed in path + VolID is a mount point (IOW, the way stage the volume). This will ensure it is the staging path, and as a bonus we can also check if the path + VolID is a file/directory to determine if it was a block based or filesystem based PV that is in use, and not resize the block PV on the node.

Reply:
kubernetes sends the stagingPath as of now, why cant the check be added in follow up PR as an enhancement? once spec 1.2 is implemented ( the code remains same) instead of volumePath we will be using stagingPath

there is a check present code to make sure its filesystem or block

Reply:
Depending on how a CO behaves is bad practice as such, what happens if this changes for some reason in Kubernetes? We will have to scramble at that point. I do not see why we need to get a patch in with a possible known issue that can be handled.

Also the ask is not an enhancement, the CSI 1.1.0 spec as such states this path can be either staging or publish.

Reply:
Also, if this is an XFS formatted PV I think this will fail to resize as the passed in volumePath is not a mount point (if it is the staging path).

Reply:
@ShyamsundarR as @Madhu-1 I dont think this is a concern at all now. Couple of reasons behind , we really dont have to worry about COs other than kubernetes and also kubernetes being the CO here always send staging path in the volume_path. Its not going to be changed to publish path for a reason that it could cause  backward compatibility breakage. Considering all these I believe this approach is good enough, if really needed , I can place a followup PR .

Reply:
Closing out the comments here,
- If volume path is the publish path, an unrelated PV can cause a resize request of another PV (say pv2) by crafting a stash file that points to pv2
    - This is not a concern as kubernetes always sends in the stage path is the current resolution
- We need to fix volume path to include the volID as otherwise XFS resize will fail [here](https://github.com/humblec/ceph-csi/blob/a25772c784e4b32be70bdc0ee848f4a7e619c685/vendor/k8s.io/kubernetes/pkg/util/resizefs/resizefs_linux.go#L57-L58).
- The detection of device [here](https://github.com/humblec/ceph-csi/blob/a25772c784e4b32be70bdc0ee848f4a7e619c685/vendor/k8s.io/kubernetes/pkg/util/resizefs/resizefs_linux.go#L40-L51) is insufficient to determine if this is a block volume, as a user of the block volume can also format the image as one of the FS types that is processed by Resize.
    - I am assuming this is also a concern for later.

We possibly need to open an issue to track all of the above, to ensure we deal with them as needed later.